CodeStarterCamp_Week4/Routine.swift

```diff
@@ -0,0 +1,106 @@
+//
+//  Routine.swift
+//  CodeStarterCamp_Week4
+//
+//  Created by QL Jeong on 2023/09/11.
+//
+
+import Foundation
+
+struct Routine {
+    let name: String
+    let activities: [Activity]
+
+    func start(with people: People) {
+        print("루틴을 몇 번 반복할까요?")
+        var count = 0
+        repeat {
+            let countResult = readRepeatCount()
+            switch countResult {
+            case .success(let value):
+                count = value
+            case .failure(let error):
+                print(errorMessage(for: error))
+            }
+        } while count < 1
+
+        print(sessionRepeat(count: count, with: people))
+    }
+}
+
+// MARK: - Private
+
+extension Routine {
+    private func readRepeatCount() -> Result<Int, RoutineInputError> {
+        guard let readValue = readLine() else {
+            return .failure(.cannotRead)
+        }
+
+        guard let result = Int(readValue) else {
+            return .failure(.notInt)
+        }
+
+        if result < 1 {
+            return .failure(.notPositiveNumber)
+        }
+
+        return .success(result)
+    }
+
+    private func sessionRepeat(count: Int, with people: People) -> String {
+        var result = ""
+        var foundError = false
+        for number in 1...count {
+            if foundError {
+                break
+            }
+
+            result += "\(number.koreanNumberString(style: .rankCardinal)) 번째 \(name)을(를) 시작합니다.\n"
+            for activity in activities {
+                if foundError {
+                    break
+                }
+
+                let workoutResult = workout(activity, with: people)
+                switch workoutResult {
+                case .success(let message):
+                    result += message
+                case .failure(let error):
+                    result += errorMessage(for: error)
+                    foundError = true
+                }
+                result += "\n"
+            }
+        }
+        result += people.conditionMessage
+
+        return result
+    }
+
+    private func workout(_ activity: Activity, with people: People) -> Result<String, PeopleWorkoutError> {
+        do {
+            return .success(try people.workout(activity))
+        }
+        catch {
+            return .failure(error as! PeopleWorkoutError)
+        }
+    }
```

Comment:
sessionRepeat 메서드에서 오류처리를 하지 않고 여기에서 Result타입으로 변환하시는 이유가 있을까요?
Result 타입의 `init(catching:)`이니셜라이저를 활용하면 조금 간단해집니다.
```suggestion
    private func workout(_ activity: Activity, with people: People) -> Result<String, Error> {
        return Result { try people.workout(activity) }
    }
```
이 정도는 따로 메서드로 분리하지 않아도 오류처리 구문으로 `sessionRepeat` 내부에서 작성해도 괜찮을 것 같습니다.

또한, 지금은 한 종류의 에러밖에 던져지지 않기 때문에 강제 다운캐스팅을 사용해도 무방하지만
다른 종류의 에러가 던져질 경우에는 위험하기 때문에 사용을 지양하는 편이 좋다고 생각합니다.

errorMessage를 출력하는 부분은 `LocalizedError` 프로토콜을 활용해주셔도 좋을 것 같습니다.

Reply:
Error 에 LocalizedError 적용 하는부분은 지난 작업에서 사용 한 적이 있음으로 사용 법은 알고 있습니다.

과제의 print 는 `피로도가 100 이상입니다. 루틴을 중단합니다.`  문구 이며
정의된 PersonWorkoutError 는 Person.workout() 에서 발생 되는 에러 입니다.
또한 에러의 내용은 `피로도가 100 이상일때 진행 하려는 경우` 가 됩니다.

여기서 문제는 LocalizedError 를 적용하면 `피로도가 100 이상이어서 진행 할 수 없다.` 가 됩니다.
하지만 `루틴을 중단합니다.` 라는 메세지를 넣어야 함으로 message 를 Routine 에서 일부러 생성 하였습니다.

`피로도가 100 이상입니다.` 와 `루틴을 중단합니다.` 를 각각 합성을 해도 될꺼 같지만
그렇게 해도 별도로 처리가 필요한건 똑같습니다.
그래서 이렇게 구현 하였습니다.

LocalizedError 적용을 해서 변경 해야 한다면 한번더 말씀 부탁 드립니다.

Reply:
LocalizedError를 반드시 활용해달라는 말씀은 아닙니다.
메세지를 만드는 부분은 errorMessage 부분이고 위 코멘트에서 중요하게 봐주시면 좋겠는 부분은
"에러처리 구문 위치"와 "강제 타입캐스팅" 입니다.

따로 workout 메서드를 routine에 정의하지 않더라도 try people.workout을 호출 가능하고
Result 타입이나 do-catch문을 활용하여 catch한 에러에 맞는 오류처리를 해주시면 좋겠다는 생각입니다.

Result타입의 경우 failure case가 any Error이기 때문에 타입캐스팅이 필요하지만 do-catch를 활용하면 에러를 특정하여 코드블록을 작성할 수 있습니다.

```swift
                do {
                    let workResult = try people.workout(activity)
                    result += workResult
                } catch PeopleWorkoutError.tiredness {
                    result += "피로도가 100 이상입니다. 루틴을 중단합니다."
                } catch {
                    print(error)
                }
```

Reply:
Routine.errorMessage() 를 제거 하였습니다.